toolkit/tools/pkg/imagecustomizerlib/liveosisobuilder.go

```diff
@@ -864,3 +894,419 @@ func createLiveOSIsoImage(buildDir, baseConfigPath string, isoConfig *imagecusto
+
+	err = b.updateGrubCfg(b.artifacts.grubCfgPath, extraCommandLine)
+	if err != nil {
+		return fmt.Errorf("failed to update grub.cfg:\n%w", err)
+	}
+
+	err = b.createIsoImage(additionalIsoFiles, outputImageDir, outputImageBase)
+	if err != nil {
+		return fmt.Errorf("failed to create iso image:\n%w", err)
+	}
+
+	return nil
+}
+
+// getSizeOnDiskInBytes
+//
+//   - given a folder, it calculates the total size in bytes of its contents.
+//
+// inputs:
+//
+//   - 'rootDir':
+//     root folder to calculate its size.
+//
+// outputs:
+//
+//   - returns the size in bytes.
+func getSizeOnDiskInBytes(rootDir string) (size uint64, err error) {
+	logger.Log.Debugf("Estimating disk size for (%s)", rootDir)
+
+	duParams := []string{"-sh", rootDir}
+	duStdout, _, err := shell.Execute("du", duParams...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to find the size of the specified folder (%s):\n%w", rootDir, err)
+	}
+
+	// parse and get count and unit
+	diskSizeRegex := regexp.MustCompile(`^(\d+)([KMGT])?.*`)
+	matches := diskSizeRegex.FindStringSubmatch(duStdout)
+	if matches == nil || len(matches) < 3 {
+		return 0, fmt.Errorf("disk size (%s) has incorrect format. Expecting: ^[0-9]+[KMGT]?\t.* (e.g. 100M, 1G)", duStdout)
+	}
+
+	numString := matches[1]
+	unit := matches[2]
+	num, err := strconv.ParseUint(numString, 0, 10)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse disk size (%s):\n%w", numString, err)
+	}
+	logger.Log.Debugf("du returned=%d %s", num, unit)
+
+	// convert to bytes
+	multiplier := uint64(1)
+	switch unit {
+	case "K":
+		multiplier = diskutils.KiB
+	case "M":
+		multiplier = diskutils.MiB
+	case "G":
+		multiplier = diskutils.GiB
+	case "T":
+		multiplier = diskutils.TiB
+	case "":
+		return 0, fmt.Errorf("disk size (%s) must have a suffix (i.e. K, M, G, or T)", numString)
+	}
+	num *= multiplier
+
+	return num, nil
+}
+
+// getDiskSizeEstimateInMBs
+//
+//   - given a folder, it calculates the size of a disk image that can hold
+//     all of its contents.
+//   - The amount of disk space a file occupies depends on the block size of the
+//     host file system. If many files are smaller than a block size, there will
+//     be a lot of waste. If files are very large, there will be very little
+//     waste. It is hard to predict how much disk space a set of a files will
+//     occupy without enumerating the sizes of all the files and knowing the
+//     target block size. In this function, we use an optimistic approach which
+//     calculates the required disk space by multiplying the total file size by
+//     a safety factor - i.e. safe that it will be able t hold all the contents.
+//
+// inputs:
+//
+//   - 'rootDir':
+//     root folder to calculate its size.
+//   - 'safetyFactor':
+//     a multiplier used with the total number of bytes calculated.
+//
+// outputs:
+//
+//   - returns the size in mega bytes.
+func getDiskSizeEstimateInMBs(rootDir string, safetyFactor uint64) (size uint64, err error) {
```

Comment:
Should safety factor be a floating point?

Reply:
fair. pending.